### PR TITLE
[RFC] Add MDIO APIs for IEEE 802.3 Clause 45

### PR DESCRIPTION
**Pull request by manuargue**
## Introduction

This RFC proposes to extend MDIO APIs to support IEEE 802.3 Clause 45 MDIO access.

### Problem description

Presently, the MDIO API offers two generic interfaces to perform read and write access over the MDIO bus. The MDIO bus controller must select the protocol supported through the [`protocol` binding property](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/mdio/mdio-controller.yaml#L11), such as Clause 22, Clause 45 or other propietary protocols.

Current design presents the following limitations:
1. The generic MDIO read/write API mixes Clause 22 and Clause 45 access.
2. Some MDIO bus controllers can support more than one protocol, and devices compliant with Clause 22 or Clause 45 specs can co-exist in the same bus. Drivers that need to support both protocols are forced to create driver-specific APIs and ignore the `protocol` binding property, such is the case of [ADIN2111 driver](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/mdio/mdio_adin2111.h).

### Proposed change

The proposed changes intend to:
1. Separate MDIO read/write APIs for Clause 22/generic access and Clause 45 access.
2. Allow the MDIO bus controller drivers to support multiple protocols through standard Zephyr APIs, if supported by hardware.

## Detailed RFC

1. Keep `mdio_read()`/`mdio_write()` generic API for Clause 22 and proprietary protocols such as Micrel SMI (which is common in Ethernet Switches and it uses a 8-bit data protocol transmitted in the 16-bit frame format of MIIM), but rename argument `devad` to `regad` to indicate this is the register address in a given PHY device address and to not be confused with the Clause 45 device address within a port.
2. Add `mdio_read_c45()`/`mdio_write_c45()` for Clause 45 access, following a similar approach to [Linux MDIO API](https://github.com/torvalds/linux/commit/4e4aafcddbbfcdd6eed5780e190fcbfac8b4685a#diff-654636a9655cea98fa84d6b8daf1f39cabaae9b10f9dfd866c24825e71de0142R367-R371) and remove the `protocol` binding property.
3. Add new MDIO header with generic opcodes, MMD and registers addresses to be used by the MDIO and PHY drivers. For now includes minimal BASE-T1/10BASE-T1L definitions used in the ADIN2111 driver.
4. Update existing MDIO drivers to implement the Clause 45 APIs whenever supported by hardware.
5. Add `read_c45`/`write_c45` MDIO shell commands.

This PR was tested with:
`west build -b <board> samples/net/telnet -- -DCONFIG_MDIO=y -DCONFIG_MDIO_SHELL=y`
for the following boards:
- `sam_e70_xplained` (build-only)
- `esp32_ethernet_kit`, plus changes described in [enabling-ethernet](https://docs.zephyrproject.org/latest/boards/xtensa/esp32_ethernet_kit/doc/index.html#enabling-ethernet) (build-only)
- a downstream board with `adi,adin2111` (see https://github.com/zephyrproject-rtos/zephyr/pull/62353#issuecomment-1737327631)
- `mr_canhubk3` board with a downstream Clause 22/Clause 45 compatible MDIO driver managing a Clause 45 PHY.

### Dependencies

No dependencies identified.

### Concerns and Unresolved Questions

None at the moment.

## Alternatives

No other alternative identified that solve the problems presented here. Feedback and alternative solutions are welcome.

### CODE REVIEW

**Comment by manuargue** [comment]
hi @rlubos @jukkar @nandojve @GeorgeCGV Could you please have a look to the propose changes? Any feedback is welcome. Thanks

File: include/zephyr/net/mdio.h [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+/** 10BASE-T1L PMA control */
+#define MDIO_PMA_B10L_CTRL		2294
+/** 10BASE-T1L PMA status */
+#define MDIO_PMA_B10L_STAT		2295
+/** 10BASE-T1L PMA link status*/
+#define MDIO_PMA_B10L_LINK_STAT		33538
+/** 10BASE-T1L PCS control */
+#define MDIO_PCS_B10L_CTRL		2278
+/** 10BASE-T1L PCS status */
+#define MDIO_PCS_B10L_STAT		2278
```
Comment (GeorgeCGV) [inline include/zephyr/net/mdio.h:169]: Please let's keep using hex for register addresses.
Values of  `MDIO_PCS_B10L_CTRL` and `MDIO_PCS_B10L_STAT` are different.

```suggestion
#define MDIO_PMA_B10L_CTRL		0x08F6U
/** 10BASE-T1L PMA status */
#define MDIO_PMA_B10L_STAT		0x08F7U
/** 10BASE-T1L PMA link status*/
#define MDIO_PMA_B10L_LINK_STAT		0x8302U
/** 10BASE-T1L PCS control */
#define MDIO_PCS_B10L_CTRL		0x08E6U
/** 10BASE-T1L PCS status */
#define MDIO_PCS_B10L_STAT		0x08E7U
```

File: include/zephyr/net/mdio.h [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+/** BASE-T1 Auto-negotiation control */
+#define MDIO_AN_T1_CTRL			512
+/** BASE-T1 Auto-negotiation status */
+#define MDIO_AN_T1_STAT			513
+/** BASE-T1 Auto-negotiation advertisement register [15:0] */
+#define MDIO_AN_T1_ADV_L		514
+/** BASE-T1 Auto-negotiation advertisement register [31:16] */
+#define MDIO_AN_T1_ADV_M		515
+/** BASE-T1 Auto-negotiation advertisement register [47:32] */
+#define MDIO_AN_T1_ADV_H		516
```
Comment (GeorgeCGV) [inline include/zephyr/net/mdio.h:112]:
```suggestion
#define MDIO_AN_T1_CTRL			0x0200U
/** BASE-T1 Auto-negotiation status */
#define MDIO_AN_T1_STAT			0x0201U
/** BASE-T1 Auto-negotiation advertisement register [15:0] */
#define MDIO_AN_T1_ADV_L		0x0202U
/** BASE-T1 Auto-negotiation advertisement register [31:16] */
#define MDIO_AN_T1_ADV_M		0x0203U
/** BASE-T1 Auto-negotiation advertisement register [47:32] */
#define MDIO_AN_T1_ADV_H		0x0204U
```

File: include/zephyr/net/mdio.h (hunk 1 above)
Comment (manuargue) [reply include/zephyr/net/mdio.h:169]: > Please let's keep using hex for register addresses.

I've used decimal values as this is the representation used in the standard and as a reference also in Linux headers (https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/mdio.h#L70).

> Values of MDIO_PCS_B10L_CTRL and MDIO_PCS_B10L_STAT are different.
Hmm I fail to see how the values you are proposing are different than the ones in the code, beside the hex and decimal representation? e.g. 0x08E6 = 2278

Comment (manuargue) [reply include/zephyr/net/mdio.h:169]: ah ok I got where's the error, `MDIO_PCS_B10L_STAT` should be `2279` (0x08E7U)

File: include/zephyr/net/mdio.h [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+/** Devices in package (2) */
+#define MDIO_DEVS2			6
+/** Control 2 */
+#define MDIO_CTRL2			7
+/** Status 2 */
+#define MDIO_STAT2			8
+/** Package identifier (1) */
+#define MDIO_PKGID1			14
+/** Package identifier (2) */
+#define MDIO_PKGID2			15
```
Comment (GeorgeCGV) [inline include/zephyr/net/mdio.h:99]:
```suggestion
#define MDIO_CTRL1			0x0000U
/** Status 1 */
#define MDIO_STAT1			0x0001U
/** Device identifier (1) */
#define MDIO_DEVID1			0x0002U
/** Device identifier (2) */
#define MDIO_DEVID2			0x0003U
/** Speed ability */
#define MDIO_SPEED			0x0004U
/** Devices in package (1) */
#define MDIO_DEVS1			0x0005U
/** Devices in package (2) */
#define MDIO_DEVS2			0x0006U
/** Control 2 */
#define MDIO_CTRL2			0x0007U
/** Status 2 */
#define MDIO_STAT2			0x0008U
/** Package identifier (1) */
#define MDIO_PKGID1			0x000EU
/** Package identifier (2) */
#define MDIO_PKGID2			0x000FU
```

File: include/zephyr/net/mdio.h [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+/** Separated PMA (3) */
+#define MDIO_MMD_SEPARATED_PMA3		10
+/** Separated PMA (4) */
+#define MDIO_MMD_SEPARATED_PMA4		11
+/** Clause 22 extension */
+#define MDIO_MMD_C22EXT			29
+/** Vendor Specific 1 */
+#define MDIO_MMD_VENDOR_SPECIFIC1	30
+/** Vendor Specific 2 */
+#define MDIO_MMD_VENDOR_SPECIFIC2	31
```
Comment (GeorgeCGV) [inline include/zephyr/net/mdio.h:75]:
```suggestion
#define MDIO_MMD_PMAPMD			0x01U
/** WAN Interface Sublayer */
#define MDIO_MMD_WIS			0x02U
/** Physical Coding Sublayer */
#define MDIO_MMD_PCS			0x03U
/** PHY Extender Sublayer */
#define MDIO_MMD_PHYXS			0x04U
/** DTE Extender Sublayer */
#define MDIO_MMD_DTEXS			0x05U
/** Transmission Convergence */
#define MDIO_MMD_TC			0x06U
/** Auto-negotiation */
#define MDIO_MMD_AN			0x07U
/** Separated PMA (1) */
#define MDIO_MMD_SEPARATED_PMA1		0x08U
/** Separated PMA (2) */
#define MDIO_MMD_SEPARATED_PMA2		0x09U
/** Separated PMA (3) */
#define MDIO_MMD_SEPARATED_PMA3		0x0AU
/** Separated PMA (4) */
#define MDIO_MMD_SEPARATED_PMA4		0x0BU
/** Clause 22 extension */
#define MDIO_MMD_C22EXT			0x1DU
/** Vendor Specific 1 */
#define MDIO_MMD_VENDOR_SPECIFIC1	0x1EU
/** Vendor Specific 2 */
#define MDIO_MMD_VENDOR_SPECIFIC2	0x1FU
```
I see that it is taken from [linux kernel mdio.h](https://docs.huihoo.com/doxygen/linux/kernel/3.7/uapi_2linux_2mdio_8h_source.html).

However, using hex helps here when navigating reference manuals of specific ICs.

Many defines are not used in Zephyr drivers. Should they be kept?

**Comment by GeorgeCGV** [comment]
I like the proposal. It covers clause separation mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/57848#issue-1707707102.

Another concern was:
> Another point relates to MDIO locking. I would like to lock the whole device while MDIO is in use by the public (i.e. shell). That allows it to prevent access while it processes frame transfer, IRQ offload, or something else where sequential access to MDIO is required.

In Linux they introduced some `lock` and `no lock` write and read methods to solve that. As of now, the ADIN driver solves that by using `enable`, and `disable` MDIO API calls. But that doesn't seem right.

Regarding dec and hex. I know that the Linux kernel defines some mdio values in decimal. That makes it a bit harder when working on drivers. As most RMs use hexadecimal. Defining values in hexadecimal helps to avoid unnecessary conversion steps in the head.

File: include/zephyr/net/mdio.h (hunk 4 above)
Comment (manuargue) [reply include/zephyr/net/mdio.h:75]: > Many defines are not used in Zephyr drivers. Should they be kept?

I tried to add the minimal currently used, but on the other hand I didn't want to partially define the generic MMD devices and registers. In this way, we set the base definitions for future drivers to use them and can be extend as they see need.

**Comment by manuargue** [comment]
> Regarding dec and hex. I know that the Linux kernel defines some mdio values in decimal. That makes it a bit harder when working on drivers. As most RMs use hexadecimal. Defining values in hexadecimal helps to avoid unnecessary conversion steps in the head. I wonder what others think on that topic.

Thanks for your review @GeorgeCGV . I've changed the registers definitions to hex, as I also see Linux headers mixes both hex and decimal, and you got a point about the RM being all in hex.

Would you be able to run tests for the ADIN2111 on this branch?

> > Another point relates to MDIO locking. I would like to lock the whole device while MDIO is in use by the public (i.e. shell). That allows it to prevent access while it processes frame transfer, IRQ offload, or something else where sequential access to MDIO is required.
> 
> In Linux they introduced some `lock` and `no lock` write and read methods to solve that. As of now, the ADIN driver solves that by using `enable`, and `disable` MDIO API calls. But that doesn't "feel" correct

I understand the purpose of enable/disable MDIO APIs is to only enable/disable the MDIO bus access (if supported by hw), and the locking is not enforced through the API but each driver needs to implement it whenever r/w access happens, e.g. through a semaphore.

**Comment by GeorgeCGV** [comment]
@manuargue thank you

> Would you be able to run tests for the ADIN2111 on this branch?

Tested it, didn't find any issues. Works.

File: include/zephyr/drivers/mdio.h [hunk 5]
```diff
@@ -54,6 +46,14 @@ __subsystem struct mdio_driver_api {
...
 	int (*write)(const struct device *dev, uint8_t prtad, uint8_t regad,
 		     uint16_t data);
+
+	/** Read data from MDIO bus using Clause 45 access */
+	int (*read_c45)(const struct device *dev, uint8_t prtad, uint8_t devad,
+			uint16_t regad, uint16_t *data);
+
+	/** Write data to MDIO bus using Clause 45 access */
+	int (*write_c45)(const struct device *dev, uint8_t prtad, uint8_t devad,
+			 uint16_t regad, uint16_t data);
```
Comment (MaureenHelm) [inline include/zephyr/drivers/mdio.h:56]: Should update the doxygen comments on the existing read/write functions to specify Clause 22 access

**Review by MaureenHelm — APPROVED** [review]
> I like the proposal. It covers clause separation mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/57848#issue-1707707102.

I like it too, though I don't have hardware to test it.

File: include/zephyr/drivers/mdio.h (hunk 5 above)
Comment (manuargue) [reply include/zephyr/drivers/mdio.h:56]: In the context of this PR, the read/write functions are not only meant for Clause 22 access but also for proprietary protocols such as Micrel SMI, which is transmitted in the same 16-bit frame format of MIIM. Therefore I haven't updated the description to not limit it to Clause 22.

**Review by GeorgeCGV — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**